isisd/isis_sr.h

```diff
@@ -0,0 +1,223 @@
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#ifndef _FRR_ISIS_SR_H
+#define _FRR_ISIS_SR_H
+
+#include "isisd/isis_tlvs.h"
+
+#include "lib/mpls.h"
+#include "lib/typesafe.h"
+
+/*
+ * Segment Routing information is transported through the following Sub-TLVs:
+ *
+ * Sub-TLV Name                         Value   TLVs
+ * ---------------------------------------------------------------------
+ * SID Label				 1
+ *
+ * Prefix Segment Identifier		 3	135, 235, 236 and 237
+ *
+ * Adjacency Segment Identifier		31	22, 23, 141, 222 and 223
+ * LAN Adjacency Segment Identifier	32	22, 23, 141, 222 and 223
+ *
+ * Segment Routing Capability		 2	242
+ * Segment Routing Algorithm		19	242
+ * Node Maximum Stack Depth (MSD)	23	242
+ *
+ * Sub-TLV definitions, serialization and de-serialization are defined
+ * in isis_tlvs.[c,h].
+ */
+
+PREDECL_RBTREE_UNIQ(tree_sr_node)
+PREDECL_RBTREE_UNIQ(tree_sr_node_prefix)
+PREDECL_RBTREE_UNIQ(tree_sr_area_prefix)
+PREDECL_RBTREE_UNIQ(tree_sr_prefix_cfg)
+
+/* SR Prefix-SID type. */
+enum sr_prefix_type {
+	ISIS_SR_PREFIX_LOCAL = 0,
+	ISIS_SR_PREFIX_REMOTE,
+};
+
+/* SR Nexthop Information. */
+struct sr_nexthop_info {
+	mpls_label_t label;
+	time_t uptime;
+};
+
+/* SR Prefix-SID. */
+struct sr_prefix {
+	/* RB-tree entries. */
+	struct tree_sr_node_prefix_item node_entry;
+	struct tree_sr_area_prefix_item area_entry;
+
+	/* IP prefix. */
+	struct prefix prefix;
+
+	/* SID value, algorithm and flags. */
+	struct isis_prefix_sid sid;
+
+	/* Local label value. */
+	mpls_label_t local_label;
+
+	/* Prefix-SID type. */
+	enum sr_prefix_type type;
+	union {
+		struct {
+			/* Information about this local Prefix-SID. */
+			struct sr_nexthop_info info;
+		} local;
+		struct {
+			/* Route associated to this remote Prefix-SID. */
+			struct isis_route_info *rinfo;
+		} remote;
+	} u;
+
+	/* Backpointer to SR node. */
+	struct sr_node *srn;
+
+	/* Flags used while the LSPDB is being parsed. */
+	uint8_t parse_flags;
+#define F_ISIS_SR_PREFIX_SID_NEW	0x01
+#define F_ISIS_SR_PREFIX_SID_MODIFIED	0x02
+#define F_ISIS_SR_PREFIX_SID_UNCHANGED	0x04
+};
+
+/* SR node. */
+struct sr_node {
+	/* RB-tree entry. */
+	struct tree_sr_node_item entry;
+
+	/* IS-IS level: ISIS_LEVEL1 or ISIS_LEVEL2. */
+	int level;
+
+	/* IS-IS node identifier. */
+	uint8_t sysid[ISIS_SYS_ID_LEN];
+
+	/* IS-IS node SRGB (Segment Routing Global Block). */
+	struct isis_srgb srgb;
+
+	/* List of Prefix-SIDs advertised by this node. */
+	struct tree_sr_node_prefix_head prefix_sids;
+
+	/* Backpointer to IS-IS area. */
+	struct isis_area *area;
+
+	/* Flags used while the LSPDB is being parsed. */
+	uint8_t parse_flags;
+#define F_ISIS_SR_NODE_NEW		0x01
+#define F_ISIS_SR_NODE_MODIFIED		0x02
+#define F_ISIS_SR_NODE_UNCHANGED	0x04
+};
```

Comment:
I don't see why we'd need to do that any of that now. If we need to collect more information to implement whatever new feature in the future, that can be done as part of the development of that feature IMO.

Reply:
Using prefix sids as "node sids" is probably the primary use case for data center fabrics, since you don't/shouldn't have any parallel links... It might be worth discussing, but focusing on prefix sids for now seems reasonable to me.

Reply:
@riw777 Adj-SIDs are supported by this code. Olivier was requesting a different way to store Adj-SIDs in order to facilitate the implementation of upcoming features like TI-LFA. This request was addressed in one of my previous updates.